### PR TITLE
Improve cycle detection while evaluating values of a constant fields 

### LINKED ISSUE #50127: An algorithm used to break/detect cycles between constant fields exhibits non-deterministic behavior and might produce unexpected diagnostics
The following unit-test succeeds today:
```
        [Fact]
        public void Test_01()
        {
            var source =
@"enum C
{
    A = F,
    B = A + C + D + E,
    C = E,
    D = B,
    E = B,
    F = G,
    G = F,
}";

            System.Collections.Immutable.ImmutableArray<Diagnostic> listOf2 = default;
            System.Collections.Immutable.ImmutableArray<Diagnostic> listOf3 = default;

            for (int i = 0; i < 100; i++)
            {
                var compilation = CreateCompilation(source);
                var list = compilation.GetDeclarationDiagnostics();

                switch (list.Length)
                {
                    case 2:
                        listOf2 = list;
                        break;
                    case 3:
                        listOf3 = list;
                        break;
                    default:
                        Assert.True(false);
                        break;
                }

                if (!listOf2.IsDefault && !listOf3.IsDefault)
                {
                    break;
                }
            }

            listOf2.Verify(
                // (4,5): error CS0110: The evaluation of the constant value for 'C.B' involves a circular definition
                //     B = A + C + D + E,
                Diagnostic(ErrorCode.ERR_CircConstValue, "B").WithArguments("C.B").WithLocation(4, 5),
                // (8,5): error CS0110: The evaluation of the constant value for 'C.F' involves a circular definition
                //     F = G,
                Diagnostic(ErrorCode.ERR_CircConstValue, "F").WithArguments("C.F").WithLocation(8, 5)
                );

            listOf3.Verify(
                // (3,5): error CS0110: The evaluation of the constant value for 'C.A' involves a circular definition
                //     A = F,
                Diagnostic(ErrorCode.ERR_CircConstValue, "A").WithArguments("C.A").WithLocation(3, 5),
                // (4,5): error CS0110: The evaluation of the constant value for 'C.B' involves a circular definition
                //     B = A + C + D + E,
                Diagnostic(ErrorCode.ERR_CircConstValue, "B").WithArguments("C.B").WithLocation(4, 5),
                // (8,5): error CS0110: The evaluation of the constant value for 'C.F' involves a circular definition
                //     F = G,
                Diagnostic(ErrorCode.ERR_CircConstValue, "F").WithArguments("C.F").WithLocation(8, 5)
                );
        }
```

Observed:
The same code sometimes produces 2 errors and sometimes 3 errors. The error that appears/disappears is
```
                // (3,5): error CS0110: The evaluation of the constant value for 'C.A' involves a circular definition
                //     A = F,
                Diagnostic(ErrorCode.ERR_CircConstValue, "A").WithArguments("C.A").WithLocation(3, 5),
```
It is also somewhat misleading because ```C.A``` is not involved into a cycle itself, it depends on a field that is involved into a cycle that is reported separately.

I believe the intent was to report the error only for fields that are part of a cycle.

### LINKED ISSUE #50163: VS 2019 IDE crash on file open
See also [AB#960718](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/960718)

_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/VS-2019-IDE-crash-on-file-open/676387)._

---
Reproducing steps:
1. Extract project from attached zip
2. Open NpType1.vbproj in IDE
3. Open file NpMessage.vb in IDE
IDE is hanging up in several seconds ...

Probably some enums or procedures is too long to VS 2019 IDE ...

---
#### Fiona Niu[MSFT] on 8/5/2019, 01:55 AM: 

<p>Thank you for taking the time to log this issue! Could you please provide more information via the Visual Studio Feedback Tool(Help -&gt; Send Feedback -&gt; Report A Problem)so that we can conduct further research?  The feedback tool  will ensure that we collect the needed information for you without worrying about what to provide (recording, dump file or ETL trace).</p>
<p>Since this issue is now marked as Need More Info, that workflow is enabled in the Feedback Tool:</p>
<pre><code>• Open Visual Studio Feedback tool.
 
• Click the banner letting you know that you have problems requesting your attention.
 
• Click this problem from the list

• Click &quot;View their request and respond&quot; from the problem details banner

• Add a comment, in the Attachments/Record: click Start Recording

• When the Steps Recorder tool appears, perform the steps that reproduce the problem.

• When you're done, choose the Stop Record button.
 
• Wait a few minutes for Visual Studio to collect and package the information that you recorded.

• Submit. You will be able to see the comment on Developer Community. For security reasons, your files come directly to us and don't appear on Developer Community.
</code></pre>
<p>For the full instructions, please see: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/how-to-report-a-problem-with-visual-studio-2017?view=vs-2017#when-further-information-is-needed-need-more-info">https://docs.microsoft.com/en-us/visualstudio/ide/how-to-report-a-problem-with-visual-studio-2017?view=vs-2017#when-further-information-is-needed-need-more-info</a> . For information about what data is collected, see <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/developer-community-privacy?view=vs-2017#data-we-collect">https://docs.microsoft.com/en-us/visualstudio/ide/developer-community-privacy?view=vs-2017#data-we-collect</a></p>
<p>We look forward to hearing from you!</p>

#### design on 8/5/2019, 02:09 AM: 

Dumps recording ...
#### design on 8/5/2019, 02:16 AM: 

Trying dumps again ...
#### Fiona Niu[MSFT] on 8/5/2019, 07:42 PM: 

<p>Thanks a lot for providing the information. We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 3/27/2020, 08:41 AM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by AlekseyTs**
- Commit 1
C#: deterministically break/detect cycles between constant fields
Fixes #50127.
- Commit 2
VB: Avoid stack overflow while evaluating value of a constant field with long dependency chain.
Fixes #50163.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Symbols/Source/ConstantEvaluationHelpers.cs [hunk 1]
```diff
@@ -233,148 +234,95 @@ private struct Node<T> where T : class
...
             {
-                var compilation = field.DeclaringCompilation;
-                if (!compilations.ContainsKey(compilation))
-                {
-                    compilations.Add(compilation, compilations.Count);
-                }
+                fieldsInvolvedInCycles = ArrayBuilder<SourceFieldSymbolWithSyntaxReference>.GetInstance(graph.Count);
+                // We sort fields that belong to the same compilation by location to process cycles in deterministic order.
+                // Relative order between compilations is not important, cycles do not cross compilation boundaries. 
+                fieldsInvolvedInCycles.AddRange(graph.Keys.GroupBy(f => f.DeclaringCompilation).
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/ConstantEvaluationHelpers.cs:277]: >(f => f.DeclaringCompilation [](start = 66, length = 28)

Consider marking static.

File: src/Compilers/CSharp/Portable/Symbols/Source/ConstantEvaluationHelpers.cs [hunk 2]
```diff
@@ -233,148 +234,95 @@ private struct Node<T> where T : class
...
+                    SelectMany(g => g.OrderByDescending((f1, f2) => g.Key.CompareSourceLocations(f1.ErrorLocation, f2.ErrorLocation))));
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/ConstantEvaluationHelpers.cs:278]: Consider marking the outer lambda static. I assume making the inner one static will be extremely complicated as it's a comparer?

File: src/Compilers/VisualBasic/Portable/Binding/ConstantFieldsInProgress.vb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+            Friend Sub Add(field As SourceFieldSymbol)
+#If DEBUG Then
+                Debug.Assert(Not _isFrozen)
+#End If
+                _builder.Add(field)
+            End Sub
+
+#If DEBUG Then
+            Friend Sub Freeze()
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Binding/ConstantFieldsInProgress.vb:57]: Consider marking the method as `Conditional` and just wrapping the assignment in the ifdefs, to simplify caller code.

File: src/Compilers/VisualBasic/Portable/Symbols/ConstantValueUtils.vb [hunk 4]
```diff
@@ -22,44 +22,31 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
+            ' if an untyped constant gets initialized with a nothing literal, it's type should be System.Object.
+            If boundValue.IsNothingLiteral Then
+                boundValueType = binder.GetSpecialType(SpecialType.System_Object, initValueSyntax, diagnostics)
             Else
-                ' Set up a binder for this part of the type.
-                Dim containingModule = field.ContainingSourceType.ContainingSourceModule
-                Dim binder As Binder = BinderBuilder.CreateBinderForType(containingModule, equalsValueOrAsNewNodeRef.SyntaxTree, field.ContainingSourceType)
-                Dim initValueSyntax As VisualBasicSyntaxNode = equalsValueOrAsNewNodeRef.GetVisualBasicSyntax()
-
-                If initValueSyntax Is Nothing Then
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Symbols/ConstantValueUtils.vb:43]: Was this case just never hit? Trying to understand why we're removing the seeming easy-outs here.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb [hunk 5]
```diff
@@ -167,11 +167,415 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
-        Friend Overrides Function GetConstantValue(inProgress As SymbolsInProgress(Of FieldSymbol)) As ConstantValue
+        ''' <param name="inProgress">>Used to detect dependencies between constant field values.</param>
+        Friend Overrides Function GetConstantValue(inProgress As ConstantFieldsInProgress) As ConstantValue
             Return Nothing
         End Function
 
+        ''' <summary>
+        ''' Helper to get a constant value of the field with cycle detection.
+        ''' Also avoids deep recursion due to references to other constant fields in the value.
+        ''' Derived types utilizing this helper should provide storage for lazily calculated
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb:178]: >  [](start = 74, length = 1)

Nit: missing an "a" or "the"

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb [hunk 6]
```diff
@@ -167,11 +167,415 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
+
+            stack.Free()
+            [set].Free()
+            Return result
+        End Function
+
+        ''' <summary>
+        ''' Should be overridden by types utilizing <see cref="GetConstantValueImpl(ConstantFieldsInProgress)"/> helper.
+        ''' </summary>
+        Protected Overridable Function GetLazyConstantTuple() As EvaluatedConstant
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb:561]: Should we prefer abstract for these helpers?

**Comment by 333fred** [comment]
Done review pass (commit 2). Just a couple of minor comments.

File: src/Compilers/VisualBasic/Portable/Symbols/ConstantValueUtils.vb (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Symbols/ConstantValueUtils.vb:43]: >Was this case just never hit? Trying to understand why we're removing the seeming easy-outs here.

I believe so. My understanding is that it is impossible to get null SyntaxNode from a non-null SyntaxReference

---
In reply to: [550661806](https://github.com/dotnet/roslyn/pull/50169#discussion_r550661806) [](ancestors = 550661806)

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb:561]: > Should we prefer abstract for these helpers?

I don't think so. These helpers are not on MethodSymbol, are used only by two derived types utilizing GetConstantValueImpl helper and aren't meant to be used for any other purpose. It is extremely unlikely that we would introduce another form of declaration for a constant field to the language. Even if that happens, the requirements are clearly documented and any test scenario would reveal that requirements aren't met. 

---
In reply to: [550684153](https://github.com/dotnet/roslyn/pull/50169#discussion_r550684153) [](ancestors = 550684153)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

**Review by 333fred — APPROVED** [review]
LGTM (commit 3)

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb [hunk 7]
```diff
@@ -167,11 +167,411 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         ''' <summary>
         ''' Gets the constant value.
         ''' </summary>
-        ''' <param name="inProgress">The previously visited const fields; used to detect cycles.</param>
-        Friend Overrides Function GetConstantValue(inProgress As SymbolsInProgress(Of FieldSymbol)) As ConstantValue
+        ''' <param name="inProgress">>Used to detect dependencies between constant field values.</param>
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb:170]: >> [](start = 37, length = 1)

Typo?

**Review by cston — APPROVED** [review]

**Pull request merged**